wire-ios-request-strategy/Sources/Request Strategies/Conversation/Actions/AddParticipantActionHandler.swift

```diff
@@ -145,21 +146,55 @@ class AddParticipantActionHandler: ActionHandler<AddParticipantAction> {
             }
 
             eventProcessor.processConversationEvents([updateEvent])
-
             action.notifyResult(.success(Void()))
 
         case 204:
             action.notifyResult(.success(Void()))
-        default:
-            if response.httpStatus == 403 {
-                // Refresh user data since this operation might have failed
-                // due to a team member being removed/deleted from the team.
-                let users: [ZMUser]? = action.userIDs.existingObjects(in: context)
-                users?.filter(\.isTeamMember).forEach({ $0.refreshData() })
+
+        case 403:
+            // Refresh user data since this operation might have failed
+            // due to a team member being removed/deleted from the team.
+            let users: [ZMUser]? = action.userIDs.existingObjects(in: context)
+            users?.filter(\.isTeamMember).forEach({ $0.refreshData() })
+
+            action.notifyResult(.failure(ConversationAddParticipantsError(response: response) ?? .unknown))
+
+        case 523:
+            guard let failure = Payload.ResponseFailure(response, decoder: decoder),
+                  failure.label == .unreachableDomains,
+                  let unreachableDomains = failure.data?.domains,
+                  let participants = action.userIDs.existingObjects(in: context) as? [ZMUser],
+                  let conversation = ZMConversation.existingObject(for: action.conversationID, in: context)
+            else {
+                return action.notifyResult(.failure(.unknown))
+            }
+
+            let users = participants.unreachable(for: unreachableDomains)
+
+            guard !users.isEmpty else {
+                return action.notifyResult(.success(Void()))
             }
+            action.notifyResult(.failure(.unreachableUsers(users)))
+
+            conversation.appendFailedToAddUsersSystemMessage(users: Set(users),
+                                                             sender: conversation.creator,
+                                                             at: conversation.lastServerTimeStamp ?? Date())
```

Comment:
I think this logic should exist one level up in the result handler. The action is to add some participants and get a success or an error. It's not clear from the caller that it also performs additional logic after notifying the result. I would argue that this logic should happen as a consequence of receiving the error result.